examples/models/llama2/aqlm/lut_kernel.cpp

```diff
@@ -0,0 +1,148 @@
+#include "lut_kernel.h"
+
+#include <numeric>
+#include <functional>
+#include <omp.h>
+
+#include <executorch/extension/kernel_util/make_boxed_from_unboxed_functor.h>
+#include <executorch/kernels/optimized/blas/CPUBlas.h>
+
+
+template<typename fp_dtype>
+void quadruple_for(
+    int num_inputs,
+    int num_input_groups, const fp_dtype* __restrict__ lut,
+    int out_features, const uint8_t* __restrict__ b_alt,
+    fp_dtype* __restrict__ output_vec
+)
+{
+    std::memset(output_vec, 0, num_inputs * out_features * sizeof(fp_dtype));
+
+    const int lut_stride = num_input_groups * 2 * 256;
+    const int b_alt_stride = 2 * out_features;
+
+    for (int input = 0; input < num_inputs; ++input) {
+        #pragma omp parallel for num_threads(4)
+        for (int j = 0; j < num_input_groups; ++j) {
+            const fp_dtype* lut_ptr = lut + input * lut_stride + j * 2 * 256;
+            const uint8_t* b_alt_ptr = b_alt + j * b_alt_stride;
+
+            for (int i = 0; i < out_features; ++i) {
+                output_vec[input * out_features + i] += lut_ptr[b_alt_ptr[i * 2]];
+                output_vec[input * out_features + i] += lut_ptr[256 + b_alt_ptr[i * 2 + 1]];
+            }
+        }
+    }
+}
+
+#if defined(__aarch64__) && defined(__ARM_NEON)
+#include <arm_neon.h>
+void row_wise_scaling_and_bias(
+    float* __restrict__ out,
+    const float* __restrict__ scales, const float* __restrict__ bias,
+    int num_input_vectors, int out_features
+) {
+    for (int j = 0; j < out_features; j += 4) {
+        float32x4_t scales_vec = vld1q_f32(scales + j);
+        float32x4_t bias_vec;
+        if (bias != nullptr){
+            bias_vec = vld1q_f32(bias + j);
+        }
+        for (int i=0; i < num_input_vectors; ++i) {
+            float32x4_t values_vec = vld1q_f32(out + i * out_features + j);
+            values_vec = vmulq_f32(values_vec, scales_vec);
+            if (bias != nullptr) {
+                values_vec = vaddq_f32(values_vec, bias_vec);
+            }
```

Comment:
you probably want to issue an FMA (https://arm-software.github.io/acle/neon_intrinsics/advsimd.html#fused-multiply-accumulate) here if bias is not nullptr. I would also recommend generating a separate function (e.g., by adding a template parameter to ignore the bias) to make sure that you get a separate kernel generated for with and without bias so you can be sure not to pay the cost of the test and branch on every iteration.

some amount of loop unrolling is also probably advisable; hopefully the compiler will do that for you, but I would recommend checking the generated assembly.